From e534c7b6584ee7db775af62ac884c6ea1bce036d Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Wed, 22 Apr 2026 18:59:27 -0500
Subject: [PATCH] winewayland: Implement support for color management protocol.

Thanks to kode54 for finding some problems with a prior version of this patch: wine-cachyos pull 6
---
 dlls/winewayland.drv/display.c         |  67 ++++++
 dlls/winewayland.drv/wayland.c         |  10 +-
 dlls/winewayland.drv/wayland_output.c  | 311 ++++++++++++++++++++++++-
 dlls/winewayland.drv/wayland_surface.c |  13 ++
 dlls/winewayland.drv/waylanddrv.h      |  25 ++
 5 files changed, 418 insertions(+), 8 deletions(-)

diff --git a/dlls/winewayland.drv/display.c b/dlls/winewayland.drv/display.c
index 65862ed4f7a..223a0ee5e72 100644
--- a/dlls/winewayland.drv/display.c
+++ b/dlls/winewayland.drv/display.c
@@ -204,8 +204,18 @@ static void wayland_add_device_source(const struct gdi_device_manager *device_ma
     device_manager->add_source(output_info->output->name, state_flags, dpi, param);
 }
 
+/* borrowed from gamescope with permission */
+static uint8_t encode_max_luminance(float nits)
+{
+    if (nits == 0.0f)
+        return 0;
+
+    return ceilf((logf(nits / 50.0f) / logf(2.0f)) * 32.0f);
+}
+
 static UINT get_edid(struct output_info *output_info, unsigned char **edid)
 {
+    const struct wayland_primaries *primaries = &output_info->output->primaries;
     struct wayland_output_mode *mode = output_info->output->current_mode;
     const char *model = output_info->output->model;
     unsigned int edid_size = 128, extensions = 0;
@@ -214,6 +224,12 @@ static UINT get_edid(struct output_info *output_info, unsigned char **edid)
     unsigned char *data, *p, c;
     char temp_model[13] = {0};
 
+    if (output_info->output->supports_hdr)
+    {
+        edid_size += 128;
+        extensions++;
+    }
+
     if (!(*edid = calloc(edid_size, sizeof(**edid))))
         return 0;
 
@@ -245,6 +261,18 @@ static UINT get_edid(struct output_info *output_info, unsigned char **edid)
     data[21] = round(mwidth / 10.0); /* cm */
     data[22] = round(mheight / 10.0); /* cm */
     data[24] = 0x6;
+    data[25] = ((primaries->r_x & 0x3) << 6) | ((primaries->r_y & 0x3) << 4) |
+               ((primaries->g_x & 0x3) << 2) | (primaries->g_y & 0x3);
+    data[26] = ((primaries->b_x & 0x3) << 6) | ((primaries->b_y & 0x3) << 4) |
+               ((primaries->w_x & 0x3) << 2) | (primaries->w_y & 0x3);
+    data[27] = (primaries->r_x & 0x3fc) >> 2;
+    data[28] = (primaries->r_y & 0x3fc) >> 2;
+    data[29] = (primaries->g_x & 0x3fc) >> 2;
+    data[30] = (primaries->g_y & 0x3fc) >> 2;
+    data[31] = (primaries->b_x & 0x3fc) >> 2;
+    data[32] = (primaries->b_y & 0x3fc) >> 2;
+    data[33] = (primaries->w_x & 0x3fc) >> 2;
+    data[34] = (primaries->w_y & 0x3fc) >> 2;
 
     for (i = 0; i < 16; ++i) data[38 + i] = 1;
 
@@ -294,6 +322,38 @@ static UINT get_edid(struct output_info *output_info, unsigned char **edid)
         c += data[i];
     data[127] = 256 - c;
 
+    p = data;
+
+    if (output_info->output->supports_hdr)
+    {
+        p += 128;
+
+        p[0] = 2;
+        p[1] = 3;
+        p[2] = 0xb;
+
+        p += 4;
+
+        p[0] = (0x7 << 5) | 0x6; /* HDR static metadata size */
+        p[1] = 6;
+
+        /* HDR static metadata block */
+
+        p[2] = 0x7; /* ST2084 | SDR | HDR */
+        p[3] = 1;
+        p[4] = encode_max_luminance(output_info->output->max_cll);
+        p[5] = encode_max_luminance(output_info->output->max_fall);
+        p[6] = 0; /* assume undefined, games often don't implement this properly */
+
+        /* reset p to beginning of the CTA block */
+        p = data + 128;
+        c = 0;
+
+        for (i = 0; i < 127; ++i)
+            c += p[i];
+        p[127] = 256 - c;
+    }
+
     return edid_size;
 }
 
@@ -301,6 +361,7 @@ static void wayland_add_device_monitor(const struct gdi_device_manager *device_m
                                        void *param, struct output_info *output_info,
                                        struct output_info *primary)
 {
+    const char *env;
     struct gdi_monitor monitor = {0};
 
     SetRect(&monitor.rc_monitor, output_info->x, output_info->y,
@@ -311,6 +372,12 @@ static void wayland_add_device_monitor(const struct gdi_device_manager *device_m
     monitor.edid_len = get_edid(output_info, &monitor.edid);
     /* We don't have a direct way to get the work area in Wayland. */
     monitor.rc_work = monitor.rc_monitor;
+    monitor.hdr_enabled = output_info->output->supports_hdr;
+
+    if ((env = getenv("DXVK_HDR")) && atoi(env))
+    {
+        monitor.hdr_enabled = TRUE;
+    }
 
     TRACE("name=%s rc_monitor=rc_work=%s\n",
           output_info->output->name, wine_dbgstr_rect(&monitor.rc_monitor));
diff --git a/dlls/winewayland.drv/wayland.c b/dlls/winewayland.drv/wayland.c
index aa73ee555a9..58597679f88 100644
--- a/dlls/winewayland.drv/wayland.c
+++ b/dlls/winewayland.drv/wayland.c
@@ -40,7 +40,10 @@ struct wayland process_wayland =
     .text_input.mutex = PTHREAD_MUTEX_INITIALIZER,
     .data_device.mutex = PTHREAD_MUTEX_INITIALIZER,
     .output_list = {&process_wayland.output_list, &process_wayland.output_list},
-    .output_mutex = PTHREAD_MUTEX_INITIALIZER
+    .output_mutex = PTHREAD_MUTEX_INITIALIZER,
+    .supports_extended_volume = FALSE,
+    .supports_pq = FALSE,
+    .supports_scrgb = FALSE
 };
 
 /**********************************************************************
@@ -232,11 +235,16 @@ static void registry_handle_global(void *data, struct wl_registry *registry,
     }
     else if (strcmp(interface, "wp_color_manager_v1") == 0)
     {
+        struct wayland_output *output;
+
         process_wayland.wp_color_manager_v1 =
             wl_registry_bind(registry, id, &wp_color_manager_v1_interface,
                              version < 3 ? version : 3);
         wp_color_manager_v1_add_listener(process_wayland.wp_color_manager_v1,
                                          &color_manager_listener, NULL);
+        /* Add image descriptions to existing outputs. */
+        wl_list_for_each(output, &process_wayland.output_list, link)
+            wayland_output_use_image_description(output);
     }
 #ifdef WL_FIXES_ACK_GLOBAL_REMOVE
     else if (strcmp(interface, "wl_fixes") == 0)
diff --git a/dlls/winewayland.drv/wayland_output.c b/dlls/winewayland.drv/wayland_output.c
index 6332247ca50..f3b0f69f7a3 100644
--- a/dlls/winewayland.drv/wayland_output.c
+++ b/dlls/winewayland.drv/wayland_output.c
@@ -35,11 +35,16 @@ WINE_DEFAULT_DEBUG_CHANNEL(waylanddrv);
 static const int32_t default_refresh = 60000;
 static uint32_t next_output_id = 0;
 
-#define WAYLAND_OUTPUT_CHANGED_MODES      0x01
-#define WAYLAND_OUTPUT_CHANGED_NAME       0x02
-#define WAYLAND_OUTPUT_CHANGED_LOGICAL_XY 0x04
-#define WAYLAND_OUTPUT_CHANGED_LOGICAL_WH 0x08
-#define WAYLAND_OUTPUT_CHANGED_GEOMETRY   0x10
+#define WAYLAND_OUTPUT_CHANGED_MODES        0x01
+#define WAYLAND_OUTPUT_CHANGED_NAME         0x02
+#define WAYLAND_OUTPUT_CHANGED_LOGICAL_XY   0x04
+#define WAYLAND_OUTPUT_CHANGED_LOGICAL_WH   0x08
+#define WAYLAND_OUTPUT_CHANGED_GEOMETRY     0x10
+#define WAYLAND_OUTPUT_CHANGED_PRIMARIES    0x20
+#define WAYLAND_OUTPUT_CHANGED_FALL         0x40
+#define WAYLAND_OUTPUT_CHANGED_CLL          0x80
+#define WAYLAND_OUTPUT_CHANGED_REF_L        0x100
+#define WAYLAND_OUTPUT_CHANGED_MAX_TARGET_L 0x200
 
 /**********************************************************************
  *          Output handling
@@ -186,6 +191,40 @@ static void wayland_output_done(struct wayland_output *output)
         output->current.logical_h = output->pending.logical_h;
     }
 
+    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_PRIMARIES)
+    {
+        output->current.primaries = output->pending.primaries;
+    }
+
+    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_FALL)
+    {
+        output->current.max_fall = output->pending.max_fall;
+    }
+
+    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_CLL)
+    {
+        output->current.max_cll = output->pending.max_cll;
+    }
+
+    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_MAX_TARGET_L)
+    {
+        output->current.max_target_lum = output->pending.max_target_lum;
+    }
+
+    if (output->pending_flags & WAYLAND_OUTPUT_CHANGED_REF_L)
+    {
+        output->current.ref_lum = output->pending.ref_lum;
+    }
+
+    output->current.supports_hdr = FALSE;
+
+    if (process_wayland.supports_extended_volume &&
+        process_wayland.supports_pq &&
+        process_wayland.supports_scrgb)
+    {
+        output->current.supports_hdr = (output->current.max_target_lum > output->current.ref_lum);
+    }
+
     output->pending_flags = 0;
 
     /* Ensure the logical dimensions have sane values. */
@@ -198,9 +237,10 @@ static void wayland_output_done(struct wayland_output *output)
 
     pthread_mutex_unlock(&process_wayland.output_mutex);
 
-    TRACE("name=%s logical=%d,%d+%dx%d\n",
+    TRACE("name=%s logical=%d,%d+%dx%d hdr=%u\n",
           output->current.name, output->current.logical_x, output->current.logical_y,
-          output->current.logical_w, output->current.logical_h);
+          output->current.logical_w, output->current.logical_h,
+          output->current.supports_hdr);
 
     RB_FOR_EACH_ENTRY(mode, &output->current.modes, struct wayland_output_mode, entry)
     {
@@ -327,6 +367,253 @@ static const struct zxdg_output_v1_listener zxdg_output_v1_listener = {
     zxdg_output_v1_handle_description,
 };
 
+static void wayland_image_description_info_v1_done(void *data,
+                                              struct wp_image_description_info_v1 *info)
+{
+    struct wayland_output *output = data;
+    wayland_output_done(output);
+}
+
+static void wayland_image_description_info_v1_icc_file(void *data,
+                                                  struct wp_image_description_info_v1 *info,
+                                                  int32_t icc, uint32_t icc_size)
+{
+}
+
+static void wayland_image_description_info_v1_primaries_named(void *data,
+				            struct wp_image_description_info_v1 *info,
+				            uint32_t primaries)
+{
+}
+
+static void wayland_image_description_info_v1_tfpower(void *data,
+				            struct wp_image_description_info_v1 *info,
+				            uint32_t power)
+{
+}
+
+static void wayland_image_description_info_v1_tfnamed(void *data,
+				            struct wp_image_description_info_v1 *info,
+				            uint32_t named)
+{
+}
+
+static void wayland_image_description_info_v1_luminance(void *data,
+                            struct wp_image_description_info_v1 *info,
+                            uint32_t min, uint32_t max, uint32_t ref)
+{
+    struct wayland_output *output = data;
+
+    TRACE("reference luminance: %u\n", ref);
+
+    output->pending.ref_lum = ref;
+    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_REF_L;
+}
+
+static void wayland_image_description_info_v1_primaries(void *data,
+                                                   struct wp_image_description_info_v1 *info,
+                                                   int32_t r_x, int32_t r_y, int32_t g_x,
+                                                   int32_t g_y, int32_t b_x, int32_t b_y,
+                                                   int32_t w_x, int32_t w_y)
+{
+    struct wayland_output *output = data;
+
+    if (!(output->pending_flags & WAYLAND_OUTPUT_CHANGED_PRIMARIES))
+    {
+#define COPY(name) output->pending.primaries.name = round((name * 1e-6) * 1024)
+        COPY(r_x);
+        COPY(r_y);
+        COPY(g_x);
+        COPY(g_y);
+        COPY(b_x);
+        COPY(b_y);
+        COPY(w_x);
+        COPY(w_y);
+#undef COPY
+
+        TRACE("primaries: {%lf, %lf, %lf, %lf, %lf, %lf, %lf, %lf}\n",
+            r_x * 1e-6, r_y * 1e-6, g_x * 1e-6, g_y * 1e-6,
+            b_x * 1e-6, b_y * 1e-6, w_x * 1e-6, w_y * 1e-6);
+
+        output->pending_flags |= WAYLAND_OUTPUT_CHANGED_PRIMARIES;
+    }
+}
+
+static void wayland_image_description_info_v1_target_primaries(void *data,
+				 struct wp_image_description_info_v1 *info,
+				 int32_t r_x, int32_t r_y,
+				 int32_t g_x, int32_t g_y,
+				 int32_t b_x, int32_t b_y,
+				 int32_t w_x, int32_t w_y)
+{
+    struct wayland_output *output = data;
+
+#define COPY(name) output->pending.primaries.name = round((name * 1e-6) * 1024)
+    COPY(r_x);
+    COPY(r_y);
+    COPY(g_x);
+    COPY(g_y);
+    COPY(b_x);
+    COPY(b_y);
+    COPY(w_x);
+    COPY(w_y);
+#undef COPY
+
+    TRACE("primaries: {%lf, %lf, %lf, %lf, %lf, %lf, %lf, %lf}\n",
+            r_x * 1e-6, r_y * 1e-6, g_x * 1e-6, g_y * 1e-6,
+            b_x * 1e-6, b_y * 1e-6, w_x * 1e-6, w_y * 1e-6);
+
+    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_PRIMARIES;
+}
+
+static void wayland_image_description_info_v1_target_luminance(void *data,
+                            struct wp_image_description_info_v1 *info,
+                            uint32_t min, uint32_t max)
+{
+    struct wayland_output *output = data;
+
+    TRACE("max target luminance: %u\n", max);
+
+    output->pending.max_target_lum = max;
+    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_MAX_TARGET_L;
+}
+
+static void wayland_image_description_info_v1_target_max_cll(void *data,
+				            struct wp_image_description_info_v1 *info,
+				            uint32_t max)
+{
+    struct wayland_output *output = data;
+
+    TRACE("Max CLL: %u\n", max);
+
+    output->pending.max_cll = max;
+    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_CLL;
+}
+
+static void wayland_image_description_info_v1_target_max_fall(void *data,
+				            struct wp_image_description_info_v1 *info,
+				            uint32_t max)
+{
+    struct wayland_output *output = data;
+    TRACE("Max FALL: %u\n", max);
+
+    output->pending.max_fall = max;
+    output->pending_flags |= WAYLAND_OUTPUT_CHANGED_FALL;
+}
+
+static const struct wp_image_description_info_v1_listener image_description_info_listener = {
+    wayland_image_description_info_v1_done,
+    wayland_image_description_info_v1_icc_file,
+    wayland_image_description_info_v1_primaries,
+    wayland_image_description_info_v1_primaries_named,
+    wayland_image_description_info_v1_tfpower,
+    wayland_image_description_info_v1_tfnamed,
+    wayland_image_description_info_v1_luminance,
+    wayland_image_description_info_v1_target_primaries,
+    wayland_image_description_info_v1_target_luminance,
+    wayland_image_description_info_v1_target_max_cll,
+    wayland_image_description_info_v1_target_max_fall
+};
+
+static void wayland_image_description_v1_failed(void *user_data,
+                    struct wp_image_description_v1 *wp_image_description_v1,
+                    uint32_t cause, const char *msg)
+{
+    struct wayland_output *output = user_data;
+    ERR("cause=%u msg=%s\n", cause, debugstr_a(msg));
+
+    wp_image_description_v1_destroy(output->wp_image_description_v1);
+    output->wp_image_description_v1 = NULL;
+}
+
+static void wayland_image_description_v1_ready2(void *user_data,
+                    struct wp_image_description_v1 *wp_image_description_v1,
+                    uint32_t identity_hi, uint32_t identity_lo)
+{
+    struct wayland_output *output = user_data;
+    TRACE("id=%#x%x\n", identity_hi, identity_lo);
+
+    output->wp_image_description_info_v1 =
+        wp_image_description_v1_get_information(
+            output->wp_image_description_v1);
+    if (!output->wp_image_description_info_v1)
+    {
+        ERR("Failed to allocate image description info object!\n");
+        return;
+    }
+    wp_image_description_info_v1_add_listener(
+        output->wp_image_description_info_v1,
+        &image_description_info_listener, output);
+}
+
+static void wayland_image_description_v1_ready(void *user_data,
+                    struct wp_image_description_v1 *wp_image_description_v1,
+                    uint32_t identity)
+{
+    wayland_image_description_v1_ready2(user_data, wp_image_description_v1, 0, identity);
+}
+
+static const struct wp_image_description_v1_listener image_description_listener = {
+    wayland_image_description_v1_failed,
+    wayland_image_description_v1_ready,
+    wayland_image_description_v1_ready2
+};
+
+static void wayland_color_management_output_image_description_changed(void *user_data,
+                struct wp_color_management_output_v1 *wp_color_management_output_v1)
+{
+    struct wayland_output *output = user_data;
+
+    if (output->wp_image_description_v1)
+    {
+        wp_image_description_v1_destroy(output->wp_image_description_v1);
+        output->wp_image_description_v1 = NULL;
+    }
+
+    if (output->wp_image_description_info_v1)
+    {
+        wp_image_description_info_v1_destroy(output->wp_image_description_info_v1);
+        output->wp_image_description_info_v1 = NULL;
+    }
+
+    wayland_output_use_image_description(output);
+}
+
+static const struct wp_color_management_output_v1_listener color_management_output_listener = {
+    wayland_color_management_output_image_description_changed
+};
+
+void wayland_output_use_image_description(struct wayland_output *output)
+{
+    if (!output->wp_color_management_output_v1)
+    {
+        output->wp_color_management_output_v1 =
+            wp_color_manager_v1_get_output(
+                        process_wayland.wp_color_manager_v1,
+                                    output->wl_output);
+        wp_color_management_output_v1_add_listener(
+            output->wp_color_management_output_v1,
+            &color_management_output_listener, output);
+        if (!output->wp_color_management_output_v1)
+        {
+            ERR("Failed to allocate color management output object!\n");
+            return;
+        }
+    }
+    if (output->wp_image_description_v1) return;
+    output->wp_image_description_v1 =
+        wp_color_management_output_v1_get_image_description(
+            output->wp_color_management_output_v1);
+    if (!output->wp_image_description_v1)
+    {
+        ERR("Failed to allocate image description object!\n");
+        return;
+    }
+    wp_image_description_v1_add_listener(
+        output->wp_image_description_v1,
+        &image_description_listener, output);
+}
+
 /**********************************************************************
  *          wayland_output_create
  *
@@ -383,6 +670,9 @@ BOOL wayland_output_create(uint32_t id, uint32_t version)
     if (process_wayland.zxdg_output_manager_v1)
         wayland_output_use_xdg_extension(output);
 
+    if (process_wayland.wp_color_manager_v1)
+        wayland_output_use_image_description(output);
+
     pthread_mutex_lock(&process_wayland.output_mutex);
     wl_list_insert(process_wayland.output_list.prev, &output->link);
     pthread_mutex_unlock(&process_wayland.output_mutex);
@@ -413,6 +703,12 @@ void wayland_output_destroy(struct wayland_output *output)
 
     wayland_output_state_deinit(&output->pending);
     wayland_output_state_deinit(&output->current);
+    if (output->wp_color_management_output_v1)
+        wp_color_management_output_v1_destroy(output->wp_color_management_output_v1);
+    if (output->wp_image_description_info_v1)
+        wp_image_description_info_v1_destroy(output->wp_image_description_info_v1);
+    if (output->wp_image_description_v1)
+        wp_image_description_v1_destroy(output->wp_image_description_v1);
     if (output->zxdg_output_v1)
         zxdg_output_v1_destroy(output->zxdg_output_v1);
     wl_output_destroy(output->wl_output);
@@ -428,6 +724,7 @@ void wayland_output_destroy(struct wayland_output *output)
  */
 void wayland_output_use_xdg_extension(struct wayland_output *output)
 {
+    if (output->zxdg_output_v1) return;
     output->zxdg_output_v1 =
         zxdg_output_manager_v1_get_xdg_output(process_wayland.zxdg_output_manager_v1,
                                               output->wl_output);
diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index aed92609beb..490d88247c1 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -81,6 +81,13 @@ static void color_manager_supported_feature(void *data, struct wp_color_manager_
 
     TRACE("feature %u\n", feature);
 
+    pthread_mutex_lock(&process_wayland.output_mutex);
+    if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_SCRGB)
+        process_wayland.supports_scrgb = TRUE;
+    else if (feature == WP_COLOR_MANAGER_V1_FEATURE_EXTENDED_TARGET_VOLUME)
+        process_wayland.supports_extended_volume = TRUE;
+    pthread_mutex_unlock(&process_wayland.output_mutex);
+
     if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_SCRGB && (desc = wp_color_manager_v1_create_windows_scrgb(manager)))
     {
         scrgb_desc = (void *)-1;
@@ -95,6 +102,12 @@ static void color_manager_supported_feature(void *data, struct wp_color_manager_
 
 static void color_manager_supported_named_tf(void *data, struct wp_color_manager_v1 *manager, uint32_t tf)
 {
+    TRACE("named tf %u\n", tf);
+
+    pthread_mutex_lock(&process_wayland.output_mutex);
+    if (tf == WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_ST2084_PQ)
+        process_wayland.supports_pq = TRUE;
+    pthread_mutex_unlock(&process_wayland.output_mutex);
 }
 
 static void color_manager_supported_primaries(void *data, struct wp_color_manager_v1 *manager, uint32_t primaries)
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index 017951842b0..e0c06f68f07 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -243,6 +243,9 @@ struct wayland
     /* Protects the output_list and the wayland_output.current states. */
     pthread_mutex_t output_mutex;
     LONG input_serial;
+    BOOL supports_pq;
+    BOOL supports_scrgb;
+    BOOL supports_extended_volume;
 };
 
 struct wayland_output_mode
@@ -253,11 +256,24 @@ struct wayland_output_mode
     int32_t refresh;
 };
 
+struct wayland_primaries
+{
+    int32_t r_x;
+    int32_t r_y;
+    int32_t g_x;
+    int32_t g_y;
+    int32_t b_x;
+    int32_t b_y;
+    int32_t w_x;
+    int32_t w_y;
+};
+
 struct wayland_output_state
 {
     int modes_count;
     struct rb_tree modes;
     struct wayland_output_mode *current_mode;
+    struct wayland_primaries primaries;
     char *name;
     char *make;
     char *model;
@@ -265,6 +281,11 @@ struct wayland_output_state
     int logical_w, logical_h;
     int physical_w, physical_h;
     int transform;
+    uint32_t max_fall;
+    uint32_t max_cll;
+    uint32_t max_target_lum;
+    uint32_t ref_lum;
+    BOOL supports_hdr;
 };
 
 struct wayland_output
@@ -272,6 +293,9 @@ struct wayland_output
     struct wl_list link;
     struct wl_output *wl_output;
     struct zxdg_output_v1 *zxdg_output_v1;
+    struct wp_image_description_v1 *wp_image_description_v1;
+    struct wp_image_description_info_v1 *wp_image_description_info_v1;
+    struct wp_color_management_output_v1 *wp_color_management_output_v1;
     uint32_t global_id;
     unsigned int pending_flags;
     struct wayland_output_state pending;
@@ -380,6 +404,7 @@ BOOL wayland_process_init(void);
 BOOL wayland_output_create(uint32_t id, uint32_t version);
 void wayland_output_destroy(struct wayland_output *output);
 void wayland_output_use_xdg_extension(struct wayland_output *output);
+void wayland_output_use_image_description(struct wayland_output *output);
 struct wayland_output *wayland_output_for_rect(const RECT *rect);
 
 /**********************************************************************
-- 
2.56.0

